src/topic/TopicMessageQuery.js

```diff
@@ -191,83 +322,109 @@ export default class TopicMessageQuery {
                     }
-                } else {
-                    const chunkInfo =
-                        /** @type {proto.IConsensusMessageChunkInfo} */ (
-                            message.chunkInfo
-                        );
-                    const initialTransactionID =
-                        /** @type {proto.ITransactionID} */ (
-                            chunkInfo.initialTransactionID
-                        );
-                    const total = /** @type {number} */ (chunkInfo.total);
-                    const transactionId =
-                        TransactionId._fromProtobuf(
-                            initialTransactionID
-                        ).toString();
-
-                    /** @type {proto.ConsensusTopicResponse[]} */
-                    let responses = [];
-
-                    const temp = list.get(transactionId);
-                    if (temp == null) {
-                        list.set(transactionId, responses);
-                    } else {
-                        responses = temp;
+                },
+                (error) => {
+                    if (this._handle != null) {
+                        this._handle.unsubscribe();
                     }
 
-                    responses.push(message);
-
-                    if (responses.length === total) {
-                        const topicMessage = TopicMessage._ofMany(responses);
+                    if (
+                        this._attempt < this._maxAttempts &&
+                        this._retryHandler(error)
+                    ) {
+                        this._attempt += 1;
 
-                        list.delete(transactionId);
-
-                        try {
-                            listener(topicMessage);
-                        } catch (error) {
-                            errorHandler(topicMessage, error);
-                        }
+                        setTimeout(() => {
+                            this._makeServerStreamRequest(client);
+                        }, 250 * 2 ** this._attempt);
                     }
                 }
-            });
+            );
+
+        if (this._handle != null) {
+            this._handle._setCall(() => cancel());
+        }
+    }
+
+    requireNotSubscribed() {
+        if (this._handle != null) {
+            throw new Error(
+                "Cannot change fields on an already subscribed query"
+            );
+        }
+    }
 
-        handle._setCall(() => cancel());
+    /**
+     * @private
+     * @param {TopicMessage} topicMessage
+     */
+    _passTopicMessage(topicMessage) {
+        try {
+            this._counter += 1;
+
+            if (this._listener != null) {
+                this._listener(topicMessage);
+            } else {
+                throw new Error("(BUG) listener is unexpectedly not set");
+            }
+
+            if (this._limit != null && this._limit.comp(this._counter)) {
+                this._completionHandler();
```

Comment:
This is incorrect. You need to hook up the completion handler to grpc's oncomplete or end signal that signifies the stream was completed successfully. This logic is wrong. It should not looked at grouped chunked messages but each individual consensus topci response. It also is complete when endTime is exceeded. But you shouldn't duplicate this logic that server already does. Just hook into the end signal.

Reply:
I completely agree that the SDK shouldn't be trying to replicate something that the mirror already does, but from my testing the gRPC library we use calls the `end` callback for errors and the completion of the stream. This was a bit confusing, but I've tested this a bit more. I've just did some more testing, and apparently I need to branch within `on("status")` rather than connecting the completion handler to `.on("end")`.